doc/api18970.md

```diff
@@ -0,0 +1,570 @@
+or interrupt context.
+
+Functions that are **reschedule** may be invoked from thread context.
+
+Functions that are **reschedule** but not **sleep** may be invoked from
+interrupt context.
+
+### sleep
+
+The sleep attribute is used on a function that can cause the invoking
+thread to sleep.
+
+Functions that are **sleep** are implicitly **reschedule**.
+
+Functions that are **sleep** may be invoked from thread context.
+
+Functions that are **sleep** may be invoked from interrupt context if
+and only if invoked in **no-wait** mode.
+
+#### Commentary
+
+This attribute is of relevance specifically when considering application
+that use only cooperative threads, because the kernel will not replace a
+running cooperative thread at a reschedule point unless that thread has
+explicitly invoked an operation that caused it to sleep.
+
+Note that this attribute does not imply the function will sleep
+unconditionally, but that the state of the system may require the
+invoking thread to suspend, wait, or invoke `k_yield()` before it can
+complete its operation.  This behavior may be mediated by **no-wait**.
+
+### no-wait
+
+The no-wait attribute is used on a function to indicate that a parameter
+to the function can force an execution path that will not cause the
+invoking thread to sleep.
+
+Functions with this attribute may be invoked from interrupt context when
+the parameter selects the no-wait path.
+
+Functions that are **no-wait** functions are implicitly **sleep**.
+
+#### Commentary
+
+The paradigmatic case of a no-wait function is a function that takes a
+timeout, to which `K_NO_WAIT` can be passed.  The semantics of this
+special timeout value are to execute the function's operation as long as
+it can be completed immediately, and to return an error code rather than
+sleep if it cannot.
+
+It is use of the no-wait feature that allows functions like
+`k_sem_take()` to be invoked from ISRs, since it is not permitted to
+sleep in interrupt context.
+
+**NOTE** That a function has a no-wait path does not imply that taking
+that path guarantees the function is synchronous.
+
+### isr-ok
+
+The isr-ok attribute is used on a function to indicate that it can be
+called from interrupt context.  The function will use `k_is_in_isr()` to
+detect its calling context and force an execution path that will not
+cause the invoking thread to sleep.
+
+This attribute is intended for **sleep** functions that may be
+indirectly invoked from interrupt context with arguments that could
+attempt to put the invoking thread to sleep, e.g. because the function
+is not **no-wait** or the parameters do not select the no-wait path.
+
+Functions that are **isr-ok** may be always be safely invoked from
+interrupt context, and will return an error if they were unable to
+fulfill their behavior in that context.
+
+### pre-kernel-ok
+
+The pre-kernel-ok attribute is used on a function that is also
+**isr-ok** to indicate that it will take reasonable steps to ensure it
```

Comment:
Looks like the term pre-kernel-ok is a new addition since the last time I looked at this.

Are there any situations in which a function might be pre-kernel-ok, but not isr-ok? I get that in practice they're similar (both "atomic context OK"), but I wonder if it would be better to leave a little wiggle room in the definition by not assuming that `pre-kernel-ok` implies `isr-ok`.

Reply:
Yes; it's there specifically because there's API in work (#21090) that specifically needs to be safe to invoke before the kernel gets started, since one of its use cases is to bring up the system clock.  However, `k_is_in_isr()` is false before the kernel gets started, so API that wants to promote itself as usable in pre-kernel initialization must do an extra check to avoid invoking API that could sleep.  (Nothing can sleep until the kernel gets past pre-kernel initialization and stops running on the interrupt stack.)

So it's not that something might be pre-kernel-ok but not isr-ok, but rather that something might be isr-ok and not pre-kernel-ok because it doesn't do the full check.

I've added "intended" here to parallel the text in **isr-ok**.

I still have no idea what "atomic context" means, and I don't think we should be using that term without a definition, even informally.

Reply:
> Yes; it's there specifically because there's API in work (#21090)

Yep, I get why you're adding this.

>> wonder if it would be better to leave a little wiggle room in the definition by not assuming that pre-kernel-ok implies isr-ok.
>
> I've added "intended" here to parallel the text in isr-ok.

I take it from this that you are not interested in this wiggle room. Perhaps this relates to your earlier concern about slippery slopes towards excessive markup, which I think I only half understand.

> I still have no idea what "atomic context" means, and I don't think we should be using that term without a definition, even informally.

I'll propose "any execution context in which it is a programming error to invoke an operation that sleeps" as an informal definition, using the terms in this PR.

Reply:
> I take it from this that you are not interested in this wiggle room.

I understood you to be saying "don't assume that pre-kernel-ok implies isr-ok". I don't see that implication in the text, and in fact it's not true.  The text did suggest pre-kernel-ok would be *used* only for explicitly isr-ok functions because you *cannot* assume that isr-ok implies pre-kernel-ok.  I've now taken that out here and will do so in #21678 so you could have a pre-kernel-ok function that's not isr-ok; hopefully that's more like what you had in mind.

> > I still have no idea what "atomic context" means....
> 
> I'll propose "any execution context in which it is a programming error to invoke an operation that sleeps" as an informal definition, using the terms in this PR.

For "atomic" I like [this definition](https://wiki.osdev.org/Atomic_operation):

> An atomic operation is an operation that will always be executed without any other process being able to read or change state that is read or changed during the operation. It is effectively executed as a single step....

There's nothing in this that prevents an atomic operation from being interrupted or sleeping as long as those interruptions don't affect the state of the operation.

What you propose above is a useful definition that covers both pre-kernel and interrupt contexts in Zephyr, but I don't see what we're gaining by using "atomic" in conjunction with "context".  I do see that the term is [used in Linux](https://lwn.net/Articles/274695/) as you'd pointed out before, but that reference seems to be an argument for not using it to mean "can't sleep".  Maybe it should be "sleep-disallowed context".

Reply:
> but I don't see what we're gaining by using "atomic" in conjunction with "context". I do see that the term is [used in Linux](https://lwn.net/Articles/274695/) as you'd pointed out before, but that reference seems to be an argument for not using it to mean "can't sleep". Maybe it should be "sleep-disallowed context".

I'm happy to stop borrowing the term "atomic context" from Linux -- which I'm only using informally and am not proposing we include in this list of terms -- if you are unhappy with it. I just use it because it's familiar shorthand from Linux, and Zephyr borrows a lot from there.

I would like to have an informal term we can use for "times when you can't sleep," since I think it's a generally useful notion, as shown by the inclusion of pre-kernel-ok and isr-ok in this list of terms.

With apologies to the Beastie Boys, we can call it "on our way to Brooklyn" mode if you want; I don't care :).

Reply:
> I understood you to be saying "don't assume that pre-kernel-ok implies isr-ok". I don't see that implication in the text, and in fact it's not true. The text did suggest pre-kernel-ok would be _used_ only for explicitly isr-ok functions because you _cannot_ assume that isr-ok implies pre-kernel-ok. I've now taken that out here and will do so in #21678 so you could have a pre-kernel-ok function that's not isr-ok; hopefully that's more like what you had in mind.

Yep, that's clearer to me now, thanks.

Reply:
Can we just call it "non-thread context" and define it as being the union of pre-kernel and interrupt context, rather than relying on the concept of "can't sleep" for its definition?  Those two plus thread context are, AFAIK, the only contexts available.

Note that in the final version I don't try to specify what "sleep" means.  That's because it becomes something like "an operation sleeps when, if it were invoked in thread context, it would cause the running thread to transition itself to a non-running state" (which could be suspended, waiting, ready-but-not-head-of-ready-queue).  Which is horrific, and probably subtly inaccurate.

Reply:
> Can we just call it "non-thread context" and define it as being the union of pre-kernel and interrupt context, rather than relying on the concept of "can't sleep" for its definition? Those two plus thread context are, AFAIK, the only contexts available.

I'm fine with using the term "non-thread context" between ourselves, but that's not exactly what I'm referring to when I say "atomic context" -- the latter applies even in thread context whenever a spinlock is held.

Reply:
Ah.  In that case I'd want "spin-locked context" as a more useful primitive concept, and you can define "atomic context" as "non-thread or spin-locked".  Though the way I undersatnd "atomic" corresponds more to "spin-locked" alone: interrupts aren't inherently atomic since they can be nested.

AFAIK, you can't sleep in an ISR because it just won't work to sleep while running on the interrupt stack.  You can sleep while holding a spin lock, but there's a chance you'll deadlock.  (ZLIs and possibly meta-IRQ threads becoming enabled break IRQ locks, so you might be able to do it in some perverse but functional way even without SMP).

Reply:
> AFAIK, you can't sleep in an ISR because it just won't work to sleep while running on the interrupt stack. You can sleep while holding a spin lock, but there's a chance you'll deadlock.

Right, which is why I used "it is a programming error" in my definition, which I believe applies to both cases.

>  In that case I'd want "spin-locked context" as a more useful primitive concept, and you can define "atomic context" as "non-thread or spin-locked".

As far as I know, that's a fine definition for Zephyr on Arm Cortex-M. :thumbsup: